File: client/src/entities/Recipe/model/recipe.ts
import { Country } from 'src/entities/Country/model/country';
import { IStep } from '../../Step/model/step';
import { User } from 'src/entities/User/model/user';
import { Category } from './recipe-filters';

export enum Difficulty {
  EASY = 'Facile',
  NORMAL = 'Moyen',
  HARD = 'Difficile',
}

export enum PriceRange {
  VERY_CHEAP = 'Very Cheap',
  CHEAP = 'Cheap',
  NORMAL = 'Normal',
  EXPENSIVE = 'Expensive',
  VERY_EXPENSIVE = 'Very Expensive',
}

export interface Recipe {
  id: number;
  name: string;
  description: string;
  imageUrl?: string;
  preparationDuration?: number;
  category?: Category;
  priceRange?: PriceRange;
  difficulty?: Difficulty;
  createdAt: string; // Assuming ISO 8601 format string
  updatedAt: string; // Assuming ISO 8601 format string
  country?: Country;
  user?: User;
  steps?: IStep[];
}

File: api/src/main/java/com/project/peps/recipe/model/Recipe.java
package com.project.peps.recipe.model;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import com.project.peps.country.model.Country;
import com.project.peps.step.model.Step;
import com.project.peps.user.model.User;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "recipes")
public class Recipe {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "recipe_id")
    private Long id;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "image_url", length = 2048)
    private String imageUrl;

    @Column(name = "preparation_duration")
    private Integer preparationDuration;

    @Enumerated(EnumType.STRING)
    @Column(name = "category")
    private Category category;

    @Enumerated(EnumType.STRING)
    @Column(name = "price_range")
    private PriceRange priceRange;

    @Enumerated(EnumType.STRING)
    @Column(name = "difficulty")
    private Difficulty difficulty;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @ManyToOne
    @JoinColumn(name = "country_id")
    private Country country;

    @ManyToOne
    @JoinColumn(name = "user_id")
    private User user;
    
    // Relation OneToMany avec Step
    @Builder.Default // Force le builder a prendre en compte l'initialisation de la lsite
    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Step> steps = new ArrayList<>(); // On initialise toujours la liste pour éviter le NullPointerException

    /**
     * Méthode utilitaire pour ajouter une étape et maintenir la cohérence des deux côtés
     */
    public void addStep(Step step) {
        steps.add(step);
        step.setRecipe(this); // ici on lie l'étape à la recette courante
    }

    /**
     * Méthode utilitaire pour retirer une étape
     */
    public void removeStep(Step step) {
        steps.remove(step);
        step.setRecipe(null);
    }

}
